File: java_solutions/1002.find-common-characters.java
/*
 * @lc app=leetcode id=1002 lang=java
 *
 * [1002] Find Common Characters
 */

class Solution {
    public List<String> commonChars(String[] words) {
        return Arrays.stream(words)
            .map(word -> word.chars().collect(() -> new int[26], (a, c) -> a[c - 'a']++, (a1, a2) -> {}))
            .reduce((a, b) -> IntStream.range(0, 26).map(i -> a[i] = Math.min(a[i], b[i])).toArray())
            .map(finalArr -> IntStream.range(0, 26)
                .boxed()
                .flatMap(i -> Collections.nCopies(finalArr[i], String.valueOf((char) (i + 'a'))).stream())
                .collect(Collectors.toList()))
            .orElse(Collections.emptyList());
    }
}

File: kotlin_solutions_1/1002.find-common-characters.kt
/* @lc app=leetcode id=1002 lang=kotlin */
class Solution {
    fun commonChars(words: Array<String>): List<String> = words.map { word -> IntArray(26).apply { word.forEach { this[it - 'a']++ } } }.reduceOrNull { a, b -> IntArray(26) { minOf(a[it], b[it]) } }?.let { finalArr -> finalArr.indices.flatMap { i -> List(finalArr[i]) { ('a' + i).toString() } } } ?: emptyList()
}